std/zlib.d

```diff
@@ -598,26 +598,152 @@ class UnCompress
 
         if (!destbufsize)
             destbufsize = to!uint(buf.length) * 2;
-        destbuf = new ubyte[zs.avail_in * 2 + destbufsize];
-        zs.next_out = destbuf.ptr;
-        zs.avail_out = to!uint(destbuf.length);
-
-        if (zs.avail_in)
-            buf = zs.next_in[0 .. zs.avail_in] ~ cast(ubyte[]) buf;
+        auto destbuf = new ubyte[destbufsize];
+        size_t destFill;
 
         zs.next_in = cast(ubyte*) buf.ptr;
         zs.avail_in = to!uint(buf.length);
 
-        err = inflate(&zs, Z_NO_FLUSH);
-        if (err != Z_STREAM_END && err != Z_OK)
+        while (true)
         {
-            GC.free(destbuf.ptr);
-            error(err);
+            auto oldAvailIn = zs.avail_in;
+
+            zs.next_out = destbuf[destFill .. $].ptr;
+            zs.avail_out = to!uint(destbuf.length - destFill);
+
+            err = inflate(&zs, Z_NO_FLUSH);
+            if (err == Z_STREAM_END)
+            {
+                inputEnded = true;
+                break;
+            }
+            else if (err != Z_OK)
+            {
+                GC.free(destbuf.ptr);
+                error(err);
+            }
+            else if (!zs.avail_in)
+                break;
+
+            /*
+                According to the zlib manual inflate() stops when either there's
+                no more data to uncompress or the output buffer is full
+                So at this point, the output buffer is too full
+            */
+
+            destFill = destbuf.length;
+
+            if (destbuf.capacity)
+            {
+                if (destbuf.length < destbuf.capacity)
+                    destbuf.length = destbuf.capacity;
+                else
+                {
+                    auto newLength = GC.extend(destbuf.ptr, destbufsize, destbufsize);
+
+                    if (newLength && destbuf.length < destbuf.capacity)
+                        destbuf.length = destbuf.capacity;
+                    else
+                        destbuf.length += destbufsize;
```

Comment:
This doesn't seem quite right. Did you mean:

```d
if (newLength)
{
    if (destbuf.length < destbuf.capacity)
        destbuf.length = destbuf.capacity;
}
else
    destbuf.length += destbufsize;
```

Reply:
No, what I've previously tried was:
```d
if (newLength)
    destbuf.length = destbuf.capacity;
else
    destbuf.length += destbufsize;
```

but that resulted in a failure with [imageformats](https://github.com/lgvz/imageformats) where `newLength` > 0, but `destbuf.capacity` didn't change, which resulted in an `OutOfMemoryError` after looping some time.
So I changed it to fall back to "simple" extending when there's no change in capacity.

Reply:
Ah, interesting, probably to do with the capacity cache ("Caching" section on https://dlang.org/d-array-article.html). Thanks for the explanation. It would be nice to document such tricky moments as comments in the source code itself.